File: 203.Remove-Linked-List-Elements.go
// https://leetcode.com/problems/remove-linked-list-elements/description/
//
// algorithms
// Easy (34.6%)
// Total Accepted:    188.8K
// Total Submissions: 545.6K
// beats 100.0% of golang submissions

package leetcode

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func removeElements(head *ListNode, val int) *ListNode {
	for head != nil && head.Val == val {
		head = head.Next
	}

	res := head
	for res != nil && res.Next != nil {
		for res.Next != nil && res.Next.Val == val {
			res.Next = res.Next.Next
		}
		res = res.Next
	}

	return head
}

File: 203.Remove-Linked-List-Elements.java
// https://leetcode.com/problems/remove-linked-list-elements/
//
// algorithms
// Easy (35.57%)
// Total Accepted:    217,102
// Total Submissions: 610,273
// beats 100.0% of java submissions


/**
 * Definition for singly-linked list. public class ListNode { int val; ListNode
 * next; ListNode(int x) { val = x; } }
 */
class Solution {
    public ListNode removeElements(ListNode head, int val) {
        ListNode tmp = new ListNode(-1);
        tmp.next = head;
        ListNode pre = tmp;

        while (head != null) {
            if (head.val == val) {
                pre.next = head.next;
            } else {
                pre = head;
            }

            head = head.next;
        }

        return tmp.next;
    }
}